app/models/dfe_sign_in_user.rb

```diff
@@ -57,9 +60,17 @@ def self.end_session!(session)
   def user_for_service
     case service
     when :claims
-      Claims::SupportUser.find_by(email:) || Claims::User.find_by(email:)
+      Claims::SupportUser.find_by(find_by_attribute) || Claims::User.find_by(find_by_attribute)
     when :placements
-      Placements::SupportUser.find_by(email:) || Placements::User.find_by(email:)
+      Placements::SupportUser.find_by(find_by_attribute) || Placements::User.find_by(find_by_attribute)
+    end
+  end
+
+  def find_by_attribute
+    if provider == "dfe"
+      "dfe_sign_in_uid='#{dfe_sign_in_uid}' OR email='#{email}'"
```

Comment:
Will this always find by uid first?

If we have a match for both `uid` and `email`, is it dependent on the user `id` - default sort order.

My trail of thinking is:

1. SQL will filter the results down to 2 records and return them sorted by default table sort order.
2. Rails will select the first record due to `find` instead of `where`.

Reply:
So these are the two sql queries that can get executed when we use dfe sign in

It will always find by uid first. So if you have two records here the first one will be the one with the UID that matches. In theory we should not have 2 or more users with same uids or emails here
```
  Claims::SupportUser Load (0.5ms)  SELECT "users".* FROM "users" WHERE "users"."type" = $1 AND (dfe_sign_in_uid='0F1D1624-4BF2-4B3C-8BE6-57B7AB22F2CF' OR email='catalin.voineag@educa
tion.gov.uk') LIMIT $2  [["type", "Claims::SupportUser"], ["LIMIT", 1]]
  ↳ (byebug):1:in `user_for_service'
nil
  Claims::User Load (0.7ms)  SELECT "users".* FROM "users" WHERE "users"."type" = $1 AND (dfe_sign_in_uid='0F1D1624-4BF2-4B3C-8BE6-57B7AB22F2CF' OR email='catalin.voineag@education.go
v.uk') LIMIT $2  [["type", "Claims::User"], ["LIMIT", 1]]

```

Reply:
I mean, let's say we have the following:

```ruby
user_1 = create(:user, uid: 123)
user_2 = create(:user, email: "123@example.com")
```

Will an oauth payload of `{ uid: 123, email: "123@example.com" }` always find `user_1` or is it dependent on the default database order? For example, do these return the same record?

```ruby
Claims::User.order(id: :asc).find_by(find_by_attribute)
Claims::User.order(id: :desc).find_by(find_by_attribute)
```

Reply:
Oh, I see what you mean. No the current query will not always return user_1. Should be fixed in this https://github.com/DFE-Digital/itt-mentor-services/pull/184/commits/812d3b46bba0af04c83bfeca3aed1d04f66bfa75